onnxruntime/core/framework/execution_providers.h

```diff
@@ -107,8 +107,25 @@ class ExecutionProviders {
 
   // maps for fast lookup of an index into exec_providers_
   std::unordered_map<std::string, size_t> provider_idx_map_;
+
+  // currently the allocator type is an implementation detail and we don't make any  behavioral choices based on it,
+  // so exclude it from the key comparison for allocator_idx_map_.
+  // we also don't expect to have two allocators with the same name, one using an arena and one not.
+  struct OrtMemoryInfoLessThanIgnoreAllocType {
+    bool operator()(const OrtMemoryInfo& lhs, const OrtMemoryInfo& rhs) const {
```

Comment:
1. Any reason to not use OrtDevice for comparison?
2. Also, it looks like storing `id` in OrtMemoryInfo seems redundant given that we already have OrtDevice that encapsulates the device id.

Reply:
1. Do you mean OrtMemoryInfo? I used a custom comparator here as it may not always be the case that you want to ignore the allocator type. 
2. Not sure it's redundant. Couldn't you have multiple different allocators using the same device, so the 'id' here isn't necessarily 1:1 with the device id?

Reply:
1. No, I mean OrtMemoryInfo contains a member called 'device'. Any reason to not consider that for the comparison.
2. The 'id' in OrtMemoryInfo was meant to represent a device id and then someone introduced OrtDevice which also contains a member called DeviceId device_id. Hence the confusion.

Reply:
As discussed, the code was copied from OrtMemoryInfo::operator< and 'device' is ignored there. Leaving potential removal of OrtMemoryInfo::id as a separate PR as it may change public interfaces.